src/hypothesis/strategies.py

```diff
@@ -982,32 +987,61 @@ def fractions(min_value=None, max_value=None, max_denominator=None):
 
     If max_denominator is not None then the absolute value of the denominator
     of any generated values is no greater than max_denominator. Note that
-    max_denominator must be at least 1.
+    max_denominator must be None or a positive integer, and the denominator
```

Comment:
> How does that [simple solution] sound?

In any case that mine doesn't cover (see below), this will avoid the filter in literally less than one in several million cases.   Will push after rebasing.

> I'm still struggling to understand precisely what problem it is the purpose of these changes to fix.

If the min_value or max_value were not integers (and we didn't enforce this, so it's reasonable to assume that fractions are supported), the generated values were not properly bounded.  In some cases, this would *also* trigger an assertion in `integers()` which I have replaced with a validation call.  Plus consequential changes to maintain our max_denominator invariant.

> The problem scenario is finding one of bounded denominator without checking every value <= n.

I think it's reasonable to bail out with an explicit error if there is no such value with denominator between n and n-1000.  Does that fit with our compatibility commitment?  I think this check is exhaustive for almost all user inputs, and not *too* expensive at very high precision.